src/GridFieldNestedForm.php

```diff
@@ -0,0 +1,435 @@
+        }
+    }
+    
+    public function getURLHandlers($gridField)
+    {
+        return [
+            'nested/$RecordID/$NestedAction' => 'handleNestedItem',
+            'toggle/$RecordID' => 'toggleNestedItem',
+            'POST movetoparent' => 'handleMoveToParent'
+        ];
+    }
+
+    /**
+     * @param GridField $field
+     */
+    public function getHTMLFragments($field)
+    {
+        if (DataObject::has_extension($field->getModelClass(), Hierarchy::class)) {
+            $field->setAttribute('data-url-movetoparent', $field->Link('movetoparent'));
+        }
+    }
+
+    public function handleMoveToParent(GridField $gridField, $request)
+    {
+        $move = $request->postVar('move');
+        /** @var DataList */
+        $list = $gridField->getList();
+        $id = isset($move['id']) ? (int) $move['id'] : null;
+        $to = isset($move['parent']) ? (int)$move['parent'] : null;
+        $parent = null;
+        if ($id) {
+            // should be possible either on parent or child grid field, or nested grid field from parent
+            $parent = $to ? $list->byID($to) : null;
+            if (!$parent
+                && $to
+                && $gridField->getForm()->getController() instanceof GridFieldNestedFormItemRequest
+                && $gridField->getForm()->getController()->getRecord()->ID == $to
+            ) {
+                $parent = $gridField->getForm()->getController()->getRecord();
+            }
+            $child = $list->byID($id);
+            if ($parent || $child || $to === 0) {
+                if (!$parent && $to) {
+                    $parent = DataList::create($gridField->getModelClass())->byID($to);
+                }
+                if (!$child) {
+                    $child = DataList::create($gridField->getModelClass())->byID($id);
+                }
+                if ($child) {
+                    if ($child->hasExtension(Hierarchy::class)) {
+                        $child->ParentID = $parent ? $parent->ID : 0;
+                    }
+                    $validationResult = $child->validate();
+                    if ($validationResult->isValid()) {
+                        if ($child->hasExtension(Versioned::class)) {
+                            $child->writeToStage(Versioned::DRAFT);
+                        } else {
+                            $child->write();
+                        }
+
+                        /** @var GridFieldOrderableRows */
+                        $orderableRows = $gridField->getConfig()->getComponentByType(GridFieldOrderableRows::class);
+                        if ($orderableRows) {
+                            $orderableRows->setImmediateUpdate(true);
+                            try {
+                                $orderableRows->handleReorder($gridField, $request);
+                            } catch (Exception $e) {
+                            }
+                        }
```

Comment:
You are completely right, this code is not good. The logic is a bit complicated here, since there are quite a few possible cases for the moving functionality, but I now at least refactored it to use early return instead of this deep control statements structure.